PackageRevision controller architecture docs (#971)
nephio-project/porch: docs/content/en/docs/5_architecture_and_components/controllers/packagerevision-controller/design.md

---
title: "Design"
type: docs
weight: 2
draft: true
description: |
  Internal design and architecture of the PackageRevision Controller.
---

## Controller Structure

> The PR controller is a standard controller-runtime reconciler. Its internal structure mirrors the reconciliation pipeline — each concern is handled by a dedicated sub-reconciler that returns early if its work is not needed:

Reviewer: link here to what a "standard controller-runtime reconciler" would be good. likely k8s docs has it

```
PackageRevisionReconciler
├── reconcileFinalizer()    — Finalizer + ownerReference management, deletion gating
├── reconcileSource()       — One-time package creation (init/clone/copy/upgrade)
├── reconcileRender()       — KRM function pipeline execution
└── reconcileLifecycle()    — Git lifecycle transitions, revision numbering
```

## CRD as Intent, Git as Content

The fundamental design decision is the separation of intent from content. The `PackageRevision` CRD in etcd is the source of truth for **what the user wants** — which lifecycle state the package should be in, how it was created, whether rendering is requested. Git is the source of truth for **what the package contains** — the actual KRM resource files.

> The controller bridges these two stores. A user sets `spec.lifecycle: Published` on the CRD; the controller transitions the package in Git to published state and updates `status` to reflect the result. This is standard Kubernetes controller semantics — spec is desired state, status is observed state.

Reviewer 2: ```suggestion
The controller bridges these two stores. When you set `spec.lifecycle: Published` on the CRD; the controller transitions the package in Git to published state and updates `status` to reflect the result. This is standard Kubernetes controller semantics — spec is desired state, status is observed state.
```

The cache provides six operations that cover the controller's needs:

- **GetPackageContent** — read package state and files from the cache
- **CreateNewDraft** — open a new draft for writing initial content
- **CreateDraftFromExisting** — open an existing package for modification (used by render)
- **CloseDraft** — commit a draft to Git
- **UpdateLifecycle** — transition a package's lifecycle state in Git
- **DeletePackage** — remove git refs (branches/tags) for a package

The controller never needs to know whether the underlying cache is CR-based or DB-based. It works identically with either implementation.

## Server-Side Apply for Status

All status updates use Server-Side Apply with distinct field managers to avoid ownership conflicts. This is important because multiple actors write to the same PackageRevision — the Repository Controller sets initial values during discovery, and the PR controller takes over during reconciliation.

Three field managers partition the status fields:

> **packagerev-controller** owns the core status: Ready condition, observedGeneration, revision number, publishedBy/At timestamps, upstream and self locks, and creationSource.

Reviewer: will likely need scrapping of all docs usage of packagerev CRD with this naming choice otherwise were gonna have confusion. (i know its on the way out with this but just a note)

**packagerev-controller-render** owns the render tracking fields: Rendered condition, renderingPrrResourceVersion, and observedPrrResourceVersion. Separating these prevents a lifecycle status update from accidentally clearing render state.

**packagerev-controller-kptfile** owns fields synced from the Kptfile after rendering: readinessGates, packageMetadata, and packageConditions. These are written to the CRD spec and status so that external controllers can read Kptfile-derived data without parsing package content.

## Concurrency-Limited Rendering

Rendering calls the function runner via gRPC, which is resource-intensive. Rather than allowing all 50 concurrent reconciles to render simultaneously, the controller uses a channel-based semaphore to bound concurrent renders to a configurable limit (default 20).

> When the semaphore is full, the reconcile doesn't block — it returns a `RequeueAfter` result and tries again after a short delay. This keeps the controller responsive and prevents it from overwhelming the function runner or exhausting gRPC connections.

Reviewer: very nice and clear

## Deletion Gating

Published packages cannot be deleted directly. This is a safety mechanism — deleting a published package from Git is destructive and irreversible. The controller enforces this through a finalizer:

When a user deletes the CRD, Kubernetes sets `deletionTimestamp` but the finalizer prevents actual removal. The controller checks the package's lifecycle:

- If the package is Published and its owner Repository still exists, the controller does nothing. The object stays in Terminating state until the user first transitions it to DeletionProposed.
- If the package is DeletionProposed (or any non-Published state), the controller cleans up Git refs and removes the finalizer, allowing Kubernetes to complete the deletion.
- If the owner Repository has been deleted (Kubernetes GC cascade), the controller allows deletion regardless of lifecycle — there's no point protecting packages whose repository is gone.

## OwnerReference to Repository

> Each PackageRevision gets an ownerReference pointing to its Repository CRD. This serves two purposes: it enables Kubernetes garbage collection (deleting a Repository cascades to all its packages), and it allows the controller to detect GC cascade during deletion gating.

Reviewer: silly question but what is (GC cascading) or GC? i know you mention it in the point above partially but could there be a bit of info or link to some?
